sled-agent/src/metrics.rs

```diff
@@ -258,3 +250,26 @@ impl MetricsManager {
         )))
     }
 }
+
+// Return the current hostname if possible.
+#[cfg(target_os = "illumos")]
+fn hostname() -> Result<String, Error> {
+    // See netdb.h
+    const MAX_LEN: usize = 256;
+    let mut out = vec![0u8; MAX_LEN + 1];
+    if unsafe {
+        libc::gethostname(out.as_mut_ptr() as *mut libc::c_char, MAX_LEN)
+    } == 0
+    {
+        // Split into subslices by NULL bytes.
+        //
+        // Safety: We've passed an extra NULL to the gethostname(2) call, so
+        // there must be at least one of these.
+        let chunk = out.split(|x| *x == 0).next().unwrap();
```

Comment:
That seems reasonable. Another option is querying the hostname length directly, with `sysconf(_SC_HOST_NAME_MAX)` or similar. I was hoping to make this infallible so that we can always produce data, but that might not be realistic.